File: JavaToKotlin/apps(kotlin)/src/main/kotlin/io/dustin/apps/board/domain/user/service/ReadUserService.java
package io.dustin.apps.board.domain.user.service;

import io.dustin.apps.common.code.BoardType;
import io.dustin.apps.board.domain.user.model.SiteUser;
import io.dustin.apps.board.domain.user.model.dto.LikeItUserDto;
import io.dustin.apps.board.domain.user.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.List;

import static io.dustin.apps.common.utils.OptionalUtils.getEntity;

@RequiredArgsConstructor // 초기화 되지않은 final 필드나, @NonNull 이 붙은 필드에 대해 생성자를 생성해 줍니다.
@Service
public class ReadUserService {

    private final UserRepository userRepository;

    public SiteUser getUser(Long userId) {
        return getEntity(userRepository.findById(userId), SiteUser.class, "site user not found");
    }

    public List<LikeItUserDto> likeItUsers(Long boardId, BoardType boardType, Pageable pageable) {
        return userRepository.findAllLikeItUser(boardId, boardType, pageable);
    }


}

File: dustin/src/main/kotlin/io/dustin/domain/user/service/ReadUserService.kt
package io.dustin.domain.user.service

import io.dustin.common.extensions.findByIdOrThrow
import io.dustin.domain.user.repository.UserRepository
import org.springframework.data.domain.Pageable
import org.springframework.data.relational.core.query.Query
import org.springframework.stereotype.Service

@Service
class ReadUserService(
    private val userRepository: UserRepository
) {

    /**
     * findById와 확장 함수로 생성한 findByIdOrThrow로 나눈 이유는 하나이다.
     * 기본적으로 해당 API를 사용할 때 전역으로 처리해야 하는 경우 외에 비지니스 로직내에서 던지지 말아야 하는 경우도 발생할 수 있다.
     */
    fun users(pageable: Pageable) = userRepository.findAllBy(pageable)

    fun userById(id: Long) = userRepository.findById(id)

    fun userByIdOrThrow(id: Long) = userRepository.findByIdOrThrow(id)

    fun totalCount() = userRepository.count()

    fun usersByQuery(match: Query) = userRepository.usersByQuery(match)

    fun totalCountByQuery(match: Query) = userRepository.totalCountByQuery(match)


}